### ISSUE TITLE
ndarray subclass selection: Add field to user-defined dtypes / dtype=object to automatically determine array type?

### ISSUE THREAD

**Issue by EricCousineau-TRI** [numpy/numpy#10887]
Follow-up from #10876, Relates #8892

For certain dtypes that have some constraints on algebra that `np.ndarray` does not normally provide, it seems like it would be convenient to teach `np.array` to possible determine the array type based on the dtypes contained in the array.

From my limited understanding, the downstream user has to explicitly ensure they use the correct wrapping class.

Example: From the example in #10876, I'd like to ensure `np.dot`, `np.matmul`, `np.equal`, etc. follow the algebra that I want. If this is best achieved by an array subclass, it'd be nice to have this automatically determined for me via `np.array` by defining a field like `__array_subclass__` on `Variable`, `Expression`, etc., rather than depending on the user to select the correct array type.

After tinkering some briefly with `pandas`, here's the existing functionality for `DatetimeIndex` + `Timestamp` (maybe not the best example based on algebra, but namely for dtypes and inference):
```
>>> import numpy as np
>>> import pandas as pd
>>> ts = pd.Timestamp('2013-01-01', freq='D')
>>> pd.DatetimeIndex([ts, ts])
DatetimeIndex(['2013-01-01', '2013-01-01'], dtype='datetime64[ns]', freq=None)
>>> np.array([ts, ts])
array([Timestamp('2013-01-01 00:00:00', freq='D'),
       Timestamp('2013-01-01 00:00:00', freq='D')], dtype=object)
```

**Comment by eric-wieser** [comment]
This seems like a bandaid for not making custom dtypes flexible enough, although that's not necessarily an argument against it.

**Comment by EricCousineau-TRI** [comment]
Agreed; if we can make `np.dot`, `np.matmul`, etc. customizable for more than uniform dtypes, then yes, this issue could certainly be closed.

That being said, it *could* be nice to add custom ufuncs / general methods to the subclassed array. An example being `AutoDiffArray.derivatives()`, which could concatenate the derivatives of all of the individual `AutoDiffScalar`s into a single matrix for simpler linear algebra.

**Comment by mattip** [comment]
Isn't this answered by `__array_ufunc__` (once `matmul` becomes a ufunc)?
